nova_vm/src/ecmascript/builtins/global_object.rs

```diff
@@ -112,15 +134,511 @@ impl BuiltinIntrinsic for GlobalObjectUnescape {
+    // 9. Let declaredFunctionNames be a new empty List.
+    let mut declared_function_names = HashSet::new();
+
+    // 10. For each element d of varDeclarations, in reverse List order, do
+    for d in var_declarations.iter().rev() {
+        // a. If d is not either a VariableDeclaration, a ForBinding, or a BindingIdentifier, then
+        if let VarScopedDeclaration::Function(d) = *d {
+            // i. Assert: d is either a FunctionDeclaration, a GeneratorDeclaration, an AsyncFunctionDeclaration, or an AsyncGeneratorDeclaration.
+            // ii. NOTE: If there are multiple function declarations for the same name, the last declaration is used.
+            // iii. Let fn be the sole element of the BoundNames of d.
+            let mut function_name = None;
+            d.bound_names(&mut |identifier| {
+                assert!(function_name.is_none());
+                function_name = Some(identifier.name.clone());
+            });
+            let function_name = function_name.unwrap();
+            // iv. If declaredFunctionNames does not contain fn, then
+            if declared_function_names.insert(function_name.clone()) {
+                // 1. If varEnv is a Global Environment Record, then
+                if let EnvironmentIndex::Global(var_env) = var_env {
+                    // a. Let fnDefinable be ? varEnv.CanDeclareGlobalFunction(fn).
+                    let function_name = String::from_str(agent, function_name.as_str());
+                    let fn_definable = var_env.can_declare_global_function(agent, function_name)?;
+
+                    // b. If fnDefinable is false, throw a TypeError exception.
+                    if !fn_definable {
+                        return Err(agent.throw_exception(
+                            ExceptionType::TypeError,
+                            "Cannot declare global function.",
+                        ));
+                    }
+                }
+
+                // 2. Append fn to declaredFunctionNames.
+                // 3. Insert d as the first element of functionsToInitialize.
+                functions_to_initialize.push(d);
+            }
+        }
+    }
+
+    // 11. Let declaredVarNames be a new empty List.
+    let mut declared_var_names = HashSet::new();
+
+    // 12. For each element d of varDeclarations, do
+    for d in var_declarations {
+        // a. If d is either a VariableDeclaration, a ForBinding, or a BindingIdentifier, then
+        if let VarScopedDeclaration::Variable(d) = d {
+            // i. For each String vn of the BoundNames of d, do
+            let mut bound_names = vec![];
+            d.id.bound_names(&mut |identifier| {
+                bound_names.push(identifier.name.clone());
+            });
+            for vn in bound_names {
+                // 1. If declaredFunctionNames does not contain vn, then
+                if !declared_function_names.contains(&vn) {
+                    let vn = String::from_str(agent, vn.as_str());
```

Comment:
**suggestion:** I think you can take ownership of `vn` and use `from_string`.

Reply:
It's a `Atom` so I think that would just do the same thing: `String::from_string(agent, vn.into_string());`

Reply:
Oh yeah, true. My bad.